File: dotnet/DesignPatterns/BehavioralDesignPatterns/CommandDesignPattern/CommandDesignPattern.Example2/RemoteControl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CommandDesignPattern.Example2;

public class RemoteControl
{
    private ICommand[] onCommands;
    private ICommand[] offCommands;
    private ICommand lastCommand;

    public RemoteControl()
    {
        onCommands = new ICommand[7]; 
        offCommands = new ICommand[7];

        // Null Object Pattern - Boş komut
        ICommand noCommand = new NoCommand();
        for (int i = 0; i < 7; i++)
        {
            onCommands[i] = noCommand;
            offCommands[i] = noCommand;
        }
        lastCommand = noCommand;
    }

    public void SetCommand(int slot, ICommand onCommand, ICommand offCommand)
    {
        onCommands[slot] = onCommand;
        offCommands[slot] = offCommand;
    }

    public void OnButtonPressed(int slot)
    {
        onCommands[slot].Execute();
        lastCommand = onCommands[slot];
    }

    public void OffButtonPressed(int slot)
    {
        offCommands[slot].Execute();
        lastCommand = offCommands[slot];
    }

    public void UndoButtonPressed()
    {
        lastCommand.Undo();
    }

    public void PrintCommands()
    {
        Console.WriteLine("\n----- Uzaktan Kumanda -----");
        for (int i = 0; i < onCommands.Length; i++)
        {
            Console.WriteLine($"[slot {i}] {onCommands[i].GetType().Name} | {offCommands[i].GetType().Name}");
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/CommandDesignPattern/CommandDesignPattern.Example2/src/RemoteControl.ts
import { ICommand } from "./ICommand";
import { NoCommand } from "./Commands";

const SLOTS = 7;

/**
 * Invoker — the remote control with on/off button pairs and single-step undo.
 */
export class RemoteControl {
  private readonly _onCommands: ICommand[];
  private readonly _offCommands: ICommand[];
  private _lastCommand: ICommand;

  constructor() {
    const noCommand = new NoCommand();
    this._onCommands  = Array.from({ length: SLOTS }, () => noCommand);
    this._offCommands = Array.from({ length: SLOTS }, () => noCommand);
    this._lastCommand = noCommand;
  }

  setCommand(slot: number, onCommand: ICommand, offCommand: ICommand): void {
    this._onCommands[slot]  = onCommand;
    this._offCommands[slot] = offCommand;
  }

  onButtonPressed(slot: number): void {
    this._onCommands[slot].execute();
    this._lastCommand = this._onCommands[slot];
  }

  offButtonPressed(slot: number): void {
    this._offCommands[slot].execute();
    this._lastCommand = this._offCommands[slot];
  }

  undoButtonPressed(): void {
    this._lastCommand.undo();
  }

  printCommands(): void {
    console.log("\n----- Remote Control -----");
    for (let i = 0; i < SLOTS; i++) {
      console.log(`[slot ${i}] ${this._onCommands[i].constructor.name} | ${this._offCommands[i].constructor.name}`);
    }
  }
}
